twoliter/embedded/imghelper

```diff
@@ -1,6 +1,150 @@
+  filename_prefix="${1:?}"
+  os_image_name="${2:?}"
+  data_image_name="${3:?}"
+  boot_image_name="${4:?}"
+  verity_image_name="${5:?}"
+  root_image_name="${6:?}"
+
+  os_image_name="${filename_prefix}"
+  data_image_name="${filename_prefix}${DATA_IMAGE_SUFFIX}"
+  boot_image_name="${filename_prefix}${BOOT_IMAGE_SUFFIX}"
+  verity_image_name="${filename_prefix}${VERITY_IMAGE_SUFFIX}"
+  root_image_name="${filename_prefix}${ROOT_IMAGE_SUFFIX}"
+}
+
+sanity_checks() {
+  local output_fmt partition_plan ovf_template uefi_secure_boot
+  output_fmt="${1:?}"
+  partition_plan="${2:?}"
+  ovf_template="${3:?}"
+  uefi_secure_boot="${4:?}"
+  case "${output_fmt}" in
+  raw | qcow2 | vmdk) ;;
+  *)
+    echo "unexpected image output format '${output_fmt}'" >&2
+    exit 1
+    ;;
+  esac
+
+  case "${partition_plan}" in
+  split | unified) ;;
+  *)
+    echo "unexpected partition plan '${partition_plan}'" >&2
+    exit 1
+    ;;
+  esac
+
+  # Fail fast if the OVF template doesn't exist, or doesn't match the layout.
+  if [ "${output_fmt}" == "vmdk" ]; then
+    if [ ! -s "${ovf_template}" ]; then
+      echo "required OVF template not found: ${ovf_template}" >&2
+      exit 1
+    fi
+
+    if [ "${partition_plan}" == "split" ]; then
+      if ! grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Missing data disk in OVF template, which is required for 'split' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${partition_plan}" == "unified" ]; then
+      if grep -Fq '{{DATA_DISK}}' "${ovf_template}"; then
+        echo "Incorrect data disk in OVF template, which is not supported for 'unified' layout." >&2
+        exit 1
+      fi
+    fi
+
+    if [ "${uefi_secure_boot}" == "yes" ]; then
+      if ! grep -Fq '{{DB_CERT_DER_HEX}}' "${ovf_template}"; then
+        echo "Missing CA certificate field in OVF template, which is required for Secure Boot support." >&2
+        exit 1
+      fi
+    fi
+  fi
+}
+
+get_data_image() {
+  local fmt input_dir filename_prefix
+  fmt="${1:?}"
+  input_dir="${2:?}"
+  filename_prefix="${3:?}"
+
+  if [[ "${fmt}" == "raw" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.img.lz4"
+  elif [[ "${fmt}" == "qcow2" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.qcow2"
+  elif [[ "${fmt}" == "vmdk" ]]; then
+    echo "${input_dir}/${filename_prefix}${DATA_IMAGE_SUFFIX}.vmdk"
+  fi
+}
+
+decompress_image() {
+  local fmt input_dir filename_prefix output_image
+  fmt="${1:?}"
+  input_dir="${2:?}"
+  filename_prefix="${3:?}"
+  output_image="${4:?}"
+
+  if [[ "${fmt}" == "raw" ]]; then
+    unlz4 -f "${input_dir}/${filename_prefix}.img.lz4" "${output_image}"
+  elif [[ "${fmt}" == "qcow2" ]]; then
+    qemu-img convert -f qcow2 -O raw \
+      "${input_dir}/${filename_prefix}.qcow2" "${output_image}"
+  fi
+}
+
+decompress_ova() {
+  local fmt input_ova filename_prefix output_image
+  fmt="${1:?}"
```

Comment:
I originally split this up because the we vend OVAs with the "friendly" name as opposed to the "versioned" name for our raw images. However, since we're going to generate missing symlinks in the end anyway, I can just create a symlink for the ova at the start.